core/src/main/scala/cats/data/Kleisli.scala

```diff
@@ -275,6 +279,20 @@ private trait KleisliApplicativeError[F[_], A, E] extends KleisliApplicative[F,
 
 }
 
+private trait KleisliFlatMap[F[_], A] extends FlatMap[Kleisli[F, A, ?]] {
+
+  implicit def FM: FlatMap[F]
+
+  def flatMap[B, C](fa: Kleisli[F, A, B])(f: B => Kleisli[F, A, C]): Kleisli[F, A, C] =
+    fa.flatMap(f)
+
+  def map[B, C](fa: Kleisli[F, A, B])(f: B => C): Kleisli[F, A, C] =
+    fa.map(f)
+
+  def tailRecM[B, C](b: B)(f: B => Kleisli[F, A, Either[B, C]]): Kleisli[F, A, C] =
+    Kleisli[F, A, C]({ a => FlatMap[F].tailRecM(b) { f(_).run(a) } })
+}
```

Comment:
Like this we have the some code twice:

- here in `KleisliFlatMap`
- and in `catsDataMonadReaderForKleisli`

Maybe we could create a `KleisliMonad` which can be extended by a `KleisliMonadReader` and a `KleisliMonadError` (similar as how it is done in `StateT`) ?